src/array/union/mod.rs

```diff
@@ -272,10 +272,7 @@ impl UnionArray {
 
     /// Returns the slot `index` as a [`Scalar`].
     pub fn value(&self, index: usize) -> Box<dyn Scalar> {
-        let type_ = self.types()[index];
-        let field = self.field(type_);
-        let index = self.field_slot(index);
-        new_scalar(field.as_ref(), index)
+        new_scalar(self, index)
```

Comment:
this will return a scalar with a `DataType` of `self`. I think we want the scalar to have the datatype of the field. As such, imo this should still do what it does, and `new_scalar` have a different logic (see comment below)

Reply:
I was going off the C++ implementation here https://github.com/apache/arrow/blob/a935c81b595d24179e115d64cda944efa93aa0e0/cpp/src/arrow/array/array_base.cc#L113, where the type of the array seems to be inherited by the scalar (rather than the field). 

There's also a test case https://github.com/apache/arrow/blob/d7fbf5260433dee08ae219e8aec34169cd610652/python/pyarrow/tests/test_scalars.py#L663 that seems to be consistent with this.

The reason passing in both the type and the union's field seemed important is to disambiguate cases where the union contains fields that have the same type. For example if we want to represent the following rust enum in Arrow:

```rust
enum E {
    UnitVariant1,
    UnitVariant2,
    IntVariant1(i64),
    IntVariant2(i64),
}
```

To retrieve the original enum from the UnionScalar, we need to be distinguish between the different fields of the same type. Presumably the different fields could have different names, so we could use that but it seemed much cleaner to have access to the typeid and fields.

Does this make sense? In general, unions seem to be an underused area in Arrow, and documentation has been light so I could wrong. Please let me know your thoughts.

Reply:
@jorgecarleitao After looking into this more, I propose we leave the `data_type` and `type_ changes` as-is and `type_` to `type_id`. This would be more expensive memory-wise, but would be consistent with the C++ library implementations and allow for better equality checks (for example the type id of two UnionScalars can be compared instead of the field name of the underlying schema). 

All the other scalar types inherit the list's data_type. If UnionScalar does the same, then that would be consistent and also provide more information for consumers of the scalar.

Reply:
I understand. What I am trying to say is that given a union with fields `f1` and `f2`, the datatype of the _value_ at position `i` should be of data type f1 or f2 depending on type. The same way that the value of a `ListArray` at position `i` contains an array of its inner field, not a `ListScalar`.

IMO this is independent of how we should represent a UnionScalar (with which I agree with you).

The scalar returned by `new_scalar(&dyn UnionScalar::new(...), index)` is a `UnionScalar` with data type equal to `DataType::Union(...)`, since `new_scalar` converts an index of an array into its scalar representation.